### PR TITLE
chore: go feature flags: use sdk functionality

### PR DESCRIPTION
**Pull request by liran2000**
## This PR

- use recently added sdk functionality instead of each provider implementation.
- [issue 405](https://github.com/open-feature/java-sdk-contrib/issues/405) - missing targeting key

### CODE REVIEW

**Comment by liran2000** [comment]
@thomaspoignant to review when available.

@toddbaert to see how some recent SDK functionality can help providers by using it instead of needing to implement for each provider.

**Review by Kavindu-Dodan — DISMISSED** [review]

File: providers/go-feature-flag/src/main/java/dev/openfeature/contrib/providers/gofeatureflag/GoFeatureFlagProvider.java
Comment (Kavindu-Dodan) [inline providers/go-feature-flag/src/main/java/dev/openfeature/contrib/providers/gofeatureflag/GoFeatureFlagProvider.java]: I think @thomaspoignant needs to approve the usage of evaluation response with error vs throwing an exception

Comment (liran2000) [reply providers/go-feature-flag/src/main/java/dev/openfeature/contrib/providers/gofeatureflag/GoFeatureFlagProvider.java]: see [this](https://github.com/open-feature/java-sdk/pull/561#discussion_r1295058795) comment by @toddbaert :

> throwing like this ends in the same result, but it's a bit easier to trace, and also simpler to write. This is how we generally recommend it's done. There is a small performance cost in serialization of the exception, but I think it's worth it.

When is ProviderNotReadyError thrown here, it should be handled by the SDK framework, ErrorCode.PROVIDER_NOT_READY and default value should be returned when evaluated via the client.

I just added a [test](https://github.com/open-feature/java-sdk-contrib/pull/403/files#diff-3d0ec86c96fe8d2f9224111448f23a4ccad41ee847a904fb50fd15921623b7faR187) step to enforce it.

As it is not always clear and trivial that throwing ProviderNotReadyError is handled by the framework, if you have any suggestion on how to make it clear in documentation/spec/provider guidelines, you are welcome to share your thoughts.

Comment (Kavindu-Dodan) [reply providers/go-feature-flag/src/main/java/dev/openfeature/contrib/providers/gofeatureflag/GoFeatureFlagProvider.java]: > As it is not always clear and trivial that throwing ProviderNotReadyError is handled by the framework..

Yeah, that's true. It's not trivial. Also, we have an error logging for exception [^1] which makes this slightly different.

I will wait for @thomaspoignant approval on this :) Otherwise, change looks good 

[^1]: https://github.com/open-feature/java-sdk/blob/v1.5.0/src/main/java/dev/openfeature/sdk/OpenFeatureClient.java#L140

File: providers/go-feature-flag/README.md
Comment (liran2000) [inline providers/go-feature-flag/README.md:26]: @thomaspoignant if you can confirm that this is correct

File: providers/go-feature-flag/src/main/java/dev/openfeature/contrib/providers/gofeatureflag/GoFeatureFlagProvider.java
Comment (thomaspoignant) [inline providers/go-feature-flag/src/main/java/dev/openfeature/contrib/providers/gofeatureflag/GoFeatureFlagProvider.java:474]: Good catch, I missed that the SDK was implementing that directly.

File: providers/go-feature-flag/src/test/java/dev/openfeature/contrib/providers/gofeatureflag/GoFeatureFlagProviderTest.java
Comment (thomaspoignant) [inline providers/go-feature-flag/src/test/java/dev/openfeature/contrib/providers/gofeatureflag/GoFeatureFlagProviderTest.java]: The PR is merged now, I guess we can refactor this.

**Review by thomaspoignant — APPROVED** [review]
Super nice contribution, thanks a lot this makes the code way more readable.

File: providers/go-feature-flag/src/main/java/dev/openfeature/contrib/providers/gofeatureflag/GoFeatureFlagProvider.java
Comment (liran2000) [reply providers/go-feature-flag/src/main/java/dev/openfeature/contrib/providers/gofeatureflag/GoFeatureFlagProvider.java:474]: It was just recently added.

**Comment by liran2000** [comment]
@toddbaert can this be merged?

**Comment by thomaspoignant** [comment]
@liran2000 I've merged it, thanks a lot again for your contribution 🙏

**Pull request merged**
